addons/sale_management/models/sale_order.py

```diff
@@ -208,6 +207,26 @@ class SaleOrderOption(models.Model):
     quantity = fields.Float('Quantity', required=True, digits='Product Unit of Measure', default=1)
     sequence = fields.Integer('Sequence', help="Gives the sequence order when displaying a list of optional products.")
 
+    def _update_price_discount(self):
+        for option in self:
+            if not option.product_id:
+                continue
+            product = option.product_id.with_context(
+                partner=option.order_id.partner_id,
+                quantity=option.quantity,
+                date=option.order_id.date_order,
+                pricelist=option.order_id.pricelist_id.id,
+                uom=option.uom_id.id,
+                fiscal_position=option.env.context.get('fiscal_position')
+            )
+            # To compute the discount a so line is created in cache
+            values = option._get_values_to_add_to_order()
+            new_sol = option.env['sale.order.line'].new(values)
+            new_sol._onchange_discount()
+            option.discount = new_sol.discount
+            if option.order_id.pricelist_id and option.order_id.partner_id:
+                option.price_unit = new_sol._get_display_price(product)
```

Comment:
```suggestion
            # To compute the discount a so line is created in cache
            values = option._get_values_to_add_to_order()
            new_sol = option.env['sale.order.line'].new(values)
            new_sol._onchange_discount()
            option.discount = new_sol.discount
            if option.order_id.pricelist_id and option.order_id.partner_id:
                product = option.product_id.with_context(
                    partner=option.order_id.partner_id,
                    quantity=option.quantity,
                    date=option.order_id.date_order,
                    pricelist=option.order_id.pricelist_id.id,
                    uom=option.uom_id.id,
                    fiscal_position=option.env.context.get('fiscal_position')
                )
                option.price_unit = new_sol._get_display_price(product)
```
Well, now that we move the code, let's be strict and define the variable only in its scope of usage :)